### PR TITLE
Implemented Combat Detection

### PR DESCRIPTION
**Pull request by jodlodi**

### CODE REVIEW

File: src/enemies/test_zombie/TestZombie.gd [hunk 1]
```diff
@@ -92,10 +92,22 @@ func take_damage(data: DamageData) -> void:
...
 
+## Function that handles aggroing and de-aggroing. Adds itself to the aggro_array if its targeting a Player.
+## Removes itself from the aggro_array otherwise.
+func aggro(tar: CharacterBody2D) -> void:
+	if tar != target:
+		target = tar
+		if target is Player:
+			EnemyUtil.aggro_array.append(self)
+		elif EnemyUtil.aggro_array.has(self):
+			EnemyUtil.aggro_array.erase(self)
```
Comment (pigslyer) [inline src/enemies/test_zombie/TestZombie.gd:107]: i don't like this direct access to the aggro array. what if we *did* switch to a dictionary? we'd need to change all these accesses to that.

File: src/autoloads/EnemyUtil.gd [hunk 2]
```diff
@@ -0,0 +1,3 @@
+extends Node
+
+var aggro_array : Array[CharacterBody2D]
```
Comment (pigslyer) [inline src/autoloads/EnemyUtil.gd:3]: why not add a pair of methods to add or remove a certain element from the aggro array? that way, it's easy for an enemy to register itself `EnemyUtil.begin_aggro(self)`.

Comment (pigslyer) [inline src/autoloads/EnemyUtil.gd:3]: while a lot of enemies will be CharacterBody2Ds, lots could also not be, for e.g. turrets. i don't see why we'd need to restrict this to just CharBodies.

**Review by pigslyer — CHANGES_REQUESTED** [review]
combat detection is a system, but should nonetheless be implemented properly.

File: src/autoloads/EnemyUtil.gd (hunk 2 above)
Comment (jodlodi) [reply src/autoloads/EnemyUtil.gd:3]: Whoops

File: src/autoloads/EnemyUtil.gd [hunk 3]
```diff
@@ -1,3 +1,13 @@
 extends Node
 
-var aggro_array : Array[CharacterBody2D]
+## Array that stores every [Node2D] that is currently aggrod on the [Player]
+var aggro_array : Array[Node2D]
+
+## Function that should be used whenever a [Node2D] aggros onto the [Player] character, so that
+## we can keep track of it.
+func begin_aggro(aggresor: Node2D):
+	aggro_array.append(aggresor)
```
Comment (pigslyer) [inline src/autoloads/EnemyUtil.gd:9]: this unconditional append could lead to a bug with how remove is appended - here is where using a dictionary would shine, as recurring adds wouldn't have the possibility of leaking. i'd appreciate *at least* an assert to stop from the same aggressor registering twice, though i don't think looping through the array once for begin aggro is that performance heavy

**Review by pigslyer — CHANGES_REQUESTED** [review]
and do it properly this time

**Review by pigslyer — APPROVED** [review]

**Pull request merged**